README.md
# Geo-ISO19139

  * My extended documentation: <http://perl.overmeer.net/CPAN/>
  * Development via GitHub: <https://github.com/markov2/perl5-Geo-ISO19139>
  * Download from CPAN: <ftp://ftp.cpan.org/pub/CPAN/authors/id/M/MA/MARKOV/>
  * Indexed from CPAN: <http://search.cpan.org/~markov/Geo-ISO19139/>
    and <https://metacpan.org/release/Geo-ISO19139>

ISO TC/211 (Geographic Information/Geomatics) is responsible for the ISO geographic information series of standards, in coordination with OpenGIS Consortium (OGC) and many other organisations and bodies. See http://www.isotc211.org

This distribution contains all definitions of ISO-19139: "Metadata - Implementation specification", which is the specification of many ISO-191xx standards. GML (ISO19136) which is released seperately in package Geo::GML, because it has also many pre-ISO versions.

## Development &rarr; Release

Important to know, is that I use an extension on POD to write the manuals.
The "raw" unprocessed version is visible on GitHub.  It will run without
problems, but does not contain manual-pages.

Releases to CPAN are different: "raw" documentation gets removed from
the code and translated into real POD and clean HTML.  This reformatting
is implemented with the OODoc distribution (A name I chose before OpenOffice
existed, sorry for the confusion)

Clone from github for the "raw" version.  For instance, when you want
to contribute a new feature.

On github, you can find the processed version for each release.  But the
better source is CPAN; to get it installed simply run:

```sh
   cpan -i Geo::ISO19139
```

## Contributing

When you want to contribute to this module, you do not need to provide
a perfect patch... actually: it is nearly impossible to create a patch
which I will merge without modification.  Usually, I need to adapt the
style of code and documentation to my own strict rules.

When you submit an extension, please contribute a set with

1. code

2. code documentation

3. regression tests in t/

**Please note:**
When you contribute in any way, you agree to transfer the copyrights to
Mark Overmeer (you will get the honors in the code and/or ChangeLog).
You also automatically agree that your contribution is released under
the same license as this project: licensed as perl itself.

## Copyright and License

This project is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.
See <http://dev.perl.org/licenses/>



ChangeLog

==== version history of Geo::ISO19139

All changes made by Mark Overmeer, unless explicitly assigned to
someone else.

version 0.12:

	Fixes:
	- fix metadata [Mohammad S Anwar]

version 0.11: Fri 19 Jan 15:51:12 CET 2018
	- do not use /bin/pwd in t/pod.t
	- repair used versions report
	- removed install dependency on Test::POD
	- upgrade schema's to v2012-07-13 of update 20070417
	- warn that the module is not actively used, but maintained
	- move to GIT and GitHUB

version 0.10: Thu Jun 12 14:29:30 CEST 2008
	initial release


lib/Geo/ISO19139.pm
# This code is part of distribution Geo::ISO19139.  Meta-POD processed with
# OODoc into POD and HTML manual-pages.  See README.md
# Copyright Mark Overmeer.  Licensed under the same terms as Perl itself.

package Geo::ISO19139;

use warnings;
use strict;

use Log::Report 'geo-iso', syntax => 'SHORT';
use XML::Compile::Cache ();
use XML::Compile::Util  qw/unpack_type pack_type/;

my %version2pkg =
 ( 2005 => 'Geo::ISO19139::2005'
 );

=chapter NAME
Geo::ISO19139 - access iso-19139 structures

=chapter SYNOPSIS
 # See Geo::ISO19139:2005

=chapter DESCRIPTION

ISO TC/211 (Geographic Information/Geomatics) is responsible for
the ISO geographic information series of standards, in coordination
with OpenGIS Consortium (OGC) and many other organisations and
bodies.  See L<http://www.isotc211.org>

This distribution contains all definitions of ISO-19139: "Metadata -
Implementation specification", which is the specification of many
ISO-191xx standards.  GML (ISO19136) which is released seperately
in package M<Geo::GML>, because it has also many pre-ISO versions.

In the future, it would be nice to extend this module into more of an
application base.  On the moment, it is just a schema container, which
enables you to read and write XML which use these schemas.

Name-spaces and Components:

  GMD ISO19115 Metadata
  GSS ISO19107 Spatial Schema, realized by GML
  GTS ISO19108 Temporal Schema
  GSR ISO19111 Spatial Reference Schema, realized by GML
  GCO ISO19118 Encoding, basic types
  GCO ISO19136 GML
  GMX Extended Geographic Metadata

B<WARNING:> I have not used this module for many years.  It probably
works (or is close to working).  You will get support.

=chapter METHODS

=section Constructors
=c_method new 'READER'|'WRITER'|'RW', %options

This C<new()> method will instantiate a M<Geo::ISO19139::2005> object,
when called with the required "version => 2005" option.  Other OPTIONS
are passed to M<Geo::ISO19139::2005::new()>.

=requires version VERSION
Only used when the object is created directly from this base-class.  It
determines which GML syntax is to be used.  Can be a VERSION like "2005"

=cut

sub new(@)
{   my $class = shift;
    my ($direction, %args) = @_;

    # having a default here cannot be maintained over the years.
    my $version = delete $args{version}
        or error __x"an explicit version is required\n";

    my $pkg     = $version2pkg{$version}
        or error __x"no implementation for version '{version}'"
             , version => $version;

    eval "require $pkg";
    $@ and die $@;

    $pkg->new($direction, %args);
}

1;


lib/Geo/ISO19139/2005.pm
# This code is part of distribution Geo::ISO19139.  Meta-POD processed with
# OODoc into POD and HTML manual-pages.  See README.md
# Copyright Mark Overmeer.  Licensed under the same terms as Perl itself.

package Geo::ISO19139::2005;
use base 'Geo::GML';

use warnings;
use strict;

use Geo::ISO19139::Util qw/:2005/;

use Log::Report 'geo-iso', syntax => 'SHORT';

use XML::Compile::Cache ();
use XML::Compile::Util  qw/unpack_type pack_type/;

my $xsd = __FILE__;
$xsd    =~ s!/2005+\.pm$!/xsd!;
my @xsd = glob("$xsd/2005/*/*.xsd");

=chapter NAME
Geo::ISO19139::2005 - access to the iso-19139 standard

=chapter SYNOPSIS

 use Geo::ISO19139;
 my $gml = Geo::ISO19139->new('READER', version => '2005');

 # see XML::Compile::Cache on how to use readers and writers
 my $data = $gml->reader("gml:GridCoverage")->($xmlmsg);
 my $xml  = $gml->writer($sometype)->($doc, $perldata);

 # or without help of the cache, XML::Compile::Schema
 my $r    = $gml->compile(READER => $sometype);
 my $data = $r->($xml);

 # overview (huge) on all defined elements
 $gml->printIndex;

=chapter DESCRIPTION
Implementation of the first ISO19139 release, initiated in 2005 but dated
2007-04-17.  The base class implements GML 3.2.1, which belongs to
the spec.  More in M<Geo::ISO19139>

=chapter METHODS

=section Constructors

=c_method new 'READER'|'WRITER'|'RW', %options

=option  version DEFAULT
=default version 2005

=option  gml_version VERSION
=default gml_version 3.2.1

=cut

sub init($)
{   my ($self, $args) = @_;
    $self->{GI9_version} = $args->{version} || '2005';
    $args->{version} = $args->{gml_version} || '3.2.1';

    my $pref = $args->{prefixes} ||= {};
    $pref->{gco} ||= NS_GCO_2005;
    $pref->{gmd} ||= NS_GMD_2005;
    $pref->{gmx} ||= NS_GMX_2005;
    $pref->{gsr} ||= NS_GSR_2005;
    $pref->{gss} ||= NS_GSS_2005;
    $pref->{gts} ||= NS_GTS_2005;

    $self->SUPER::init($args);
    $self->importDefinitions(\@xsd);
    $self;
}

#-------------

=section Accessors

=method gmlVersion
=method version
=cut

sub gmlVersion() {shift->SUPER::version}
sub version()    {shift->{GI9_version}}

1;


lib/Geo/ISO19139/Util.pm
# This code is part of distribution Geo::ISO19139.  Meta-POD processed with
# OODoc into POD and HTML manual-pages.  See README.md
# Copyright Mark Overmeer.  Licensed under the same terms as Perl itself.

package Geo::ISO19139::Util;
use base 'Exporter';

use warnings;
use strict;

our @r2005   = qw/NS_GCO_2005 NS_GMD_2005 NS_GMX_2005
  NS_GSR_2005 NS_GSS_2005 NS_GTS_2005 NS_GML_2005/;

our @EXPORT_OK = ( @r2005 );

our %EXPORT_TAGS =
  ( 2005 => \@r2005
  );

=chapter NAME
Geo::ISO19139::Util - iso-19139 useful constants

=chapter SYNOPSIS
=chapter DESCRIPTION
=chapter FUNCTIONS

=section Constants

=subsection Export Tags :2005
=cut

use constant NS_GML_2005 => 'http://www.opengis.net/gml/3.2';
use constant NS_GCO_2005 => 'http://www.isotc211.org/2005/gco';
use constant NS_GMD_2005 => 'http://www.isotc211.org/2005/gmd';
use constant NS_GMX_2005 => 'http://www.isotc211.org/2005/gmx';
use constant NS_GSR_2005 => 'http://www.isotc211.org/2005/gsr';
use constant NS_GSS_2005 => 'http://www.isotc211.org/2005/gss';
use constant NS_GTS_2005 => 'http://www.isotc211.org/2005/gts';

1;


Makefile.PL
use ExtUtils::MakeMaker;

use 5.010;

WriteMakefile
  ( NAME	 	 => 'Geo::ISO19139'
  , VERSION	     => '0.12'
  , PREREQ_PM    =>
   { XML::Compile         => 0.84
   , XML::Compile::Tester => 0.02
   , XML::Compile::Cache  => 0.10
   , Log::Report          => 0.19
   , Geo::GML             => 0.18

   , Test::More           => 0.54
   }

  , EXE_FILES    => []
  , AUTHOR       => 'Mark Overmeer'
  , ABSTRACT     => 'Geography metadata standard'
  , LICENSE      => 'perl_5'

  , META_MERGE =>
      { 'meta-spec' => { version => 2 }
      , resources  =>
          { repository =>
              { type => 'git'
              , url  => 'https://github.com/markov2/perl5-Geo-ISO19139.git'
              , web  => 'https://github.com/markov2/perl5-Geo-ISO19139'
              }
          , homepage => 'http://perl.overmeer.net/CPAN/'
          , license  => [ 'http://dev.perl.org/licenses/' ]
          }
      }
  );

### used by oodist during production of distribution
sub MY::postamble { <<'__POSTAMBLE' }

# for DIST
RAWDIR          = ../public_html/geo-iso19139/raw
DISTDIR         = ../public_html/geo-iso19139/source
LICENSE         = perl

# for POD
FIRST_YEAR      = 2008,2018
EMAIL           = markov@cpan.org
WEBSITE         = http://perl.overmeer.net/CPAN/
__POSTAMBLE
